# week6
python script for generating symlinks for RNAseq, ATACseq, and DNAseq data included in this repo.

The script works similarly for ATACseq and DNAseq files, by first writing functions for moving up and downt through directories, then creating a top level directory with the first system argument. The program then moves into this newly created directory to begin the symlink creation. Additionally, a dictionary was created as a reference to create rename the symlinks, with the Barcode, genotype, tissue, and biological replicate as the naming convention. The script loops through every file in the original ATACseq directory, looking for the barcode name and the Read number (R1/R2) in each file, then appends the respective genotype/tissue/bioreplicate to the string name, and finally creates a symlink using this created name to the original file.
For RNAseq, since the file hierarchy is different, the files are walked through using the os.walk function, and the undetermined indices are removed, and the readme file is translated into a pandas dataframe. The respective SampleNumber is pulled from the data frame, and the Full Sample Name is respectively taken to rename the symlink. Again if R1 or R2 is detected as the read number, then it is added into the name of the symlink. The naming convention for the RNAseq symlink is FullSampleName_ReadNumber.fq.gz and all the symlinks are placed into the same directory.

